add  developer onboarding flow

notagilbert opened the pull request:
## Overview

This PR implements a comprehensive developer onboarding system that enables developers to create professional profiles, showcase their projects, and offer services to potential clients visiting project pages.

## High-Level Architecture

The implementation creates a new onboarding domain alongside existing payment/funding features. While existing tables like `project` and `user_repository` handle repository ecosystem metadata and payment permissions, the new onboarding tables manage professional developer relationships and service offerings.

## Core Functionality: Project Maintainer Discovery

The primary business value is enabling `/projects/{org}/{repo}` pages to display:
- **Maintainers & Contributors**: Developers working on the project with their roles
- **Available Services**: What each developer offers (bug fixes, features, consulting)  
- **Pricing & Availability**: Hourly rates, response times, weekly commitment
- **Contact Information**: How to hire them for services

This transforms project pages from static information into active marketplaces for developer services.

## Technical Implementation Details

### **New Domain Tables**
The onboarding system requires 6 new tables for developer professional relationships, which existing tables cannot accommodate due to different business domains and data requirements.

### **Token Security Enhancement**  
GitHub OAuth tokens are now encrypted using AES-256-GCM with environment-based key management, replacing plain text storage for security compliance.

### **Service Type Integration**
Service categories are aligned with existing `ServiceType` enums (Support/Development/Operation/Advisory) ensuring consistency between payment and onboarding systems.

## Developer Onboarding Flow

1. **Profile Creation**: Developer accepts terms and creates profile
2. **Project Association**: Links GitHub repositories or manual projects with role definitions
3. **Income Preferences**: Selects royalties, services, or donations preference
4. **Availability Setup**: Sets weekly hours, rates, and opportunity preferences
5. **Service Offerings**: Defines specific services with pricing and response times
6. **Completion**: Profile becomes discoverable on project pages

## Database Schema

### **Core Tables**:
- `developer_profile`: User profile with terms acceptance
- `developer_project`: GitHub/manual projects with roles (creator, maintainer, contributor)
- `developer_service`: Service offerings with pricing and categories
- `service_project`: Many-to-many service-to-project associations

### **Supporting Tables**:
- `developer_income_preference`: Income type selection
- `developer_availability`: Hours, rates, opportunity preferences  
- `service_category`: Pre-seeded categories matching existing ServiceType enums

## Key Use Case: Project Pages with Maintainer Information

The onboarding tables enable a critical feature: **displaying maintainer/contributor information on project pages**.

### **Example Query for `/projects/facebook/react`:**

```sql
-- Get all developers working on facebook/react with their service offerings
-- This shows how ONE repository can have MULTIPLE developers with different roles
SELECT 
    dp_profile.name,
    dp_profile.email,
    dp_project.role,                    -- Each developer has their own role
    dp_project.merge_rights,            -- Each developer has different merge rights
    ds.service_name,
    ds.hourly_rate,
    ds.response_time_hours,
    sc.name as service_category
FROM developer_project dp_project
JOIN developer_profile dp_profile ON dp_project.developer_profile_id = dp_profile.id
JOIN developer_service ds ON ds.developer_profile_id = dp_profile.id
JOIN service_project sp ON sp.developer_service_id = ds.id 
    AND sp.developer_project_id = dp_project.id
JOIN service_category sc ON ds.service_category_id = sc.id
WHERE dp_project.project_type = 'github'
  AND dp_project.github_org = 'facebook'
  AND dp_project.github_repo = 'react'
  AND dp_profile.onboarding_completed = true;

-- Example result:
-- John Smith    | maintainer      | full_rights      | Code Review | $150/hr
-- Sarah Johnson | core_developer  | specific_areas   | Bug Fixes   | $120/hr  
-- Mike Chen     | contributor     | no_rights        | Testing     | $80/hr
```

### **Why existing user_repository cannot support this:**

```sql
-- ❌ Current user_repository only has payment permissions:
SELECT 
    au.name,
    ur.repository_user_role,  -- Only 'admin' or 'read' (not developer roles)
    ur.dow_rate              -- Payment rate (not service offerings)
FROM user_repository ur
JOIN app_user au ON ur.user_id = au.id
WHERE ur.github_owner_login = 'facebook' 
  AND ur.github_repository_name = 'react';

-- Missing: developer role, merge rights, service offerings, availability
```

## Two-Table Architecture: Payment vs Professional

The design maintains **clear separation** between two different concerns:

### **user_repository: Payment Domain**
- **Purpose**: Who can access repositories and at what cost
- **Data**: Payment permissions, DOW rates, access levels
- **Query**: "Who has paid access to this repo and what did they pay?"

### **developer_project: Professional Domain**  
- **Purpose**: Who works on projects and what services they offer
- **Data**: Developer roles, merge rights, service offerings
- **Query**: "Who maintains this project and what services do they provide?"

### **Connection Point: Project Pages**
Both tables can reference the same GitHub repository but serve different UI sections:

```typescript
// /projects/facebook/react page data
const projectData = {
  // From user_repository: Paid access section
  paidUsers: await getUserRepositoryData('facebook', 'react'),
  
  // From developer_project: Maintainer/contributor section  
  developers: await getDeveloperProjectData('facebook', 'react'),
  
  // Combined view for comprehensive project overview
  projectOverview: {
    maintainers: developers.filter(d => d.role === 'maintainer'),
    contributors: developers.filter(d => d.role === 'core_developer'),
    services: developers.flatMap(d => d.services)
  }
};
```

## API Endpoints

**Profile Management**: `POST/GET/PUT /api/v1/onboarding/profile`
**Project Linking**: `POST/PUT/DELETE /api/v1/onboarding/projects`  
**GitHub Integration**: `GET /api/v1/onboarding/github/organizations/{org}/repositories`
**Service Setup**: `POST/PUT/DELETE /api/v1/onboarding/services`
**Completion**: `POST /api/v1/onboarding/complete`

## Security & Performance

- **Encrypted Tokens**: AES-256-GCM encryption for GitHub OAuth tokens
- **Input Validation**: Comprehensive Joi schemas for all endpoints
- **Database Indexes**: Optimized for project discovery queries
- **Session Auth**: Leverages existing authentication system

## Testing

- **Migration Guide**: `MIGRATION_GUIDE.md` with encryption setup
- **API Testing**: Updated `API-Testing-Guide.md` with all endpoints
- **Unit Tests**: Repository and validation layer coverage

## Impact Summary

✅ **New Feature**: Developer marketplace on project pages  
✅ **Security**: Encrypted GitHub token storage  
✅ **Integration**: Service categories match existing ServiceType enums  
✅ **Zero Breaking Changes**: Existing functionality unaffected  
✅ **Clean Architecture**: Separate domains with clear boundaries

notagilbert committed:
wip: add developer onboarding flow

LaurianeOSE commented on src/db/migration/12.sql:
```diff
@@ -0,0 +1,128 @@
+-- Migration 12: Developer profile, projects, income preferences, availability, and services
+
+-- Developer onboarding profile
+CREATE TABLE IF NOT EXISTS developer_profile (
```
There is a lot of duplication between this table and app_user. 

1. I think we need to remove those duplicates (name, email, github_username)

2. Then, terms_accepted needs to be moved to `app_user` (because every user needs to accept the terms). 

3. Then, only remains "onboarding_completed", so I am wondering if the `developer_profile` table should not be just the`app_user` table, where we add a column `onboarding_completed`, WDYT?

notagilbert replied:
You're right about the duplicates! I've moved `terms_accepted` to `app_user` since all users need it, but kept `onboarding_completed` in `developer_profile` since only developers go through this onboarding. The table now serves as both a developer marker and tracks their onboarding status. and i make changes i might eliminate the table completely but i will leave it as it is now.

  Does this approach make sense?

LaurianeOSE replied:
Yes, let's see, I am also not sure if we need a `developer_profile` table or not, so let's leave it like that for now, and see what other requirements we will need on the enterprise side - and we will decide based on that. 

(I am working on what is the onboarding process for an enterprise user)

notagilbert committed:
wip: add missing DTOs

LaurianeOSE commented on src/db/migration/12.sql:
```diff
@@ -0,0 +1,128 @@
+);
+
+-- Developer projects (both GitHub and manual)
+CREATE TABLE IF NOT EXISTS developer_project (
```
Here, when a URL is given, it could be: 

1. A GitHub URL, in that case, we will parse the `github_org` and `github_repo`. So in that case, this is the info in the table `project`. 
2. a non-GitHub URL -> and outside of the developer enboarding process - I am not sure if we are going to not support them, and if so, how. 


So I am wondering it is not better to have the following schema (I am really not sure)

1. `github_project` that is the current `project`
2. `other_project` with the `project_url` (and `project_name` ?)

So, either we have two tables, or we update the `project` table, WDYT?

notagilbert replied:
@LaurianeOSE I think we can handle both cases with the existing developer_project table by adding a `project_id` reference to the canonical `project` table. For GitHub URLs, we'll parse `org/repo`, auto-create/find the canonical project entry, and link via `project_id` - this eliminates duplication since all devs referencing "apache/kafka" share the same project. For non-GitHub URLs, we store them directly in `project_url` with platform detection via URL parsing (supporting GitLab, Bitbucket, etc.). 

This single-table approach is simpler for queries while giving us canonical project benefits for GitHub and flexibility for other platforms. The existing schema just needs a nullable `project_id` column added.

Let me know if this affects the enterprise onboarding flow you are working on.

LaurianeOSE replied:
Let's have a quick call about that if you dont mind

Just a note: I think that I do not know all the platforms where open source can be hosted - I am sure that there are obscure edge cases. So, even if we can do some parsing for the common platforms like GitLab or Bitbucket, but we will need a fallback in any case

LaurianeOSE commented on src/db/migration/12.sql:
```diff
@@ -0,0 +1,128 @@
+);
+
+-- Developer projects (both GitHub and manual)
+CREATE TABLE IF NOT EXISTS developer_project (
```
Then, we can have multiple developers working on the same project. I think we need to have that reflect the one-to-many relation. That is the role of this table (with an updated name)

LaurianeOSE commented on src/db/migration/12.sql:
```diff
@@ -0,0 +1,128 @@
+    
+    -- GitHub project fields
+    github_org VARCHAR(255),
+    github_repo VARCHAR(255),
```
Could you keep the naming convention for that data?

See in project for example

```
    github_owner_id            BIGINT        NOT NULL,
    github_owner_login         VARCHAR(255)  NOT NULL,

    github_repository_id       BIGINT,
    github_repository_name     VARCHAR(255),
```

PS: As discussed, I do not think we need the 4 fields - but constituency helps with the refactoring we need to do for only account the github `id` and not the github names/login

LaurianeOSE commented on src/db/migration/12.sql:
```diff
@@ -0,0 +1,128 @@
+    project_url VARCHAR(500),
+    
+    -- Common fields
+    role VARCHAR(50) NOT NULL CHECK (role IN ('creator_founder', 'project_lead', 'core_developer', 'maintainer')),
```
A dev can have several roles. 

Sorry, I forgot to tell you - I mentioned it to Suné - but it's not meant to be in the front-end yet.

LaurianeOSE commented on src/db/migration/12.sql:
```diff
@@ -0,0 +1,128 @@
+    weekly_commitment INTEGER NOT NULL CHECK (weekly_commitment > 0),
+    larger_opportunities VARCHAR(10) NOT NULL CHECK (larger_opportunities IN ('yes', 'maybe', 'no')),
+    hourly_rate DECIMAL(10,2) NOT NULL CHECK (hourly_rate > 0),
+    currency VARCHAR(10) NOT NULL DEFAULT 'USD',
```
Somewhere, we need a list of supported currencies. 
So we can either have a separate table for supported currencies (but a little too overkilled IMO) 
So we just hardcode the list (see class `Currency`) and we make the check on the backend, more risky but easier.

LaurianeOSE commented on src/db/migration/12.sql:
```diff
@@ -0,0 +1,128 @@
+    id UUID PRIMARY KEY NOT NULL DEFAULT gen_random_uuid(),
+    developer_profile_id UUID NOT NULL,
+    service_category_id UUID NOT NULL,
+    service_name VARCHAR(200),
```
That is already in the table `service_category `, please remove

LaurianeOSE commented on src/db/migration/14.sql:
```diff
@@ -0,0 +1,24 @@
+
+-- Add GitHub OAuth token fields to existing github_owner table
+-- NOTE: Tokens are stored encrypted using AES-256-GCM
+ALTER TABLE github_owner 
```
I guess thoses changes should go to the `app_user` table. 

`github_owner` is to reflect the data that we have from the GitHub API. `app_user` is the one used for GitHub OAuth

LaurianeOSE replied:
Still the case

LaurianeOSE commented on src/db/onboarding/DeveloperProject.repository.ts:
```diff
@@ -0,0 +1,181 @@
+import { pool } from "../../dbPool";
+import { logger } from "../../config";
+
+export function getDeveloperProjectRepository(): DeveloperProjectRepository {
```
So, you did not want to introduce ORMs at the end?

LaurianeOSE commented on src/api/dto/onboarding/SetIncomePreference.dto.ts:
```diff
@@ -0,0 +1,3 @@
+export interface SetIncomePreferenceDto {
+    incomeType: 'royalties' | 'services' | 'donations';
```
Code style: could you use a project enum? Thnaks

LaurianeOSE commented on src/controllers/onboarding.controller.ts:
```diff
@@ -0,0 +1,656 @@
+const githubService = getGitHubService();
+
+export interface OnboardingController {
+  createProfile(req: Request, res: Response): Promise<void>;
```
Code Style: Could you please add full type hints to the method signature(s), similar to our other controllers? While it's a bit more verbose, strong typing makes the codebase more robust, easier to maintain, and simplifies debugging in the long run.

LaurianeOSE commented on src/services/github.service.ts:
```diff
@@ -24,6 +28,33 @@ export interface GitHubApi {
+  getUserRepositories(accessToken: string): Promise<Repository[]>;
+}
+
+export interface Organization {
```
Already exist, called `Owner`

LaurianeOSE commented on src/services/github.service.ts:
```diff
@@ -24,6 +28,33 @@ export interface GitHubApi {
+  description: string | null;
+}
+
+export interface GitHubRepository {
```
Already exists too, `Repository`, please update this class, instead of re-creating one

notagilbert committed:
wip: refactor developer_profile table to remove duplication with app_user

notagilbert committed:
wip: update typeScript interfaces and API endpoints for refactored developer_profile

LaurianeOSE committed:
wip

LaurianeOSE committed:
Should be ok

LaurianeOSE committed:
Should be ok

LaurianeOSE committed:
forgot to add this file

LaurianeOSE committed:
FIX

notagilbert committed:
Update DB schema/structure for onboarding and API improvements

- Add unique constraints to prevent duplicate developer rights per project/role
- Add data integrity constraints for realistic hourly rates and values
- Remove test endpoint with hardcoded user ID for security
- Implement health check endpoints for monitoring and status
- Add detailed health checks with database connectivity verification

LaurianeOSE committed:
Change file names

LaurianeOSE commented on src/db/migration/migration.ts:
```diff
@@ -21,17 +21,29 @@ export class Migration {
+      "11.2.sql",  // Project polymorphic structure
+      "12.sql",    // Developer onboarding tables
+      "13.sql",    // No-op migration (bug fix)
+      "14.sql",    // GitHub OAuth token storage
```
No 15.sql here - useful to add only for the tests

LaurianeOSE commented on .env:
```diff
@@ -27,4 +27,6 @@ STRIPE_SECRET_KEY=sk_test_51PyDEVP1HLKJuuAxQHPUd7mHZg7J67zKufLj4jBNyq2YC57XPNzTk
\ No newline at end of file
+POSTMARK_SENDER_EMAIL=lauriane@open-source-economy.com
+
+TOKEN_ENCRYPTION_KEY="r5Mnuj5G4IKkCuQsHG5bmKMfyQuGaH/clXKeyCQoNPg="
```
Could you send me the script to generate one?

notagilbert replied:
You can run this command `node scripts/generate-encryption-key.js`

LaurianeOSE commented on src/routes/v1/health.route.ts:
```diff
@@ -0,0 +1,12 @@
+const router = Router();
+
+// Basic health check endpoint (no authentication required)
+router.get("/", HealthController.healthCheck);
```
What is the plan for this endpoint? Probably Vercel can support and intergation with it?

notagilbert replied:
I was using that endpoint mainly for quick testing when running migrations, since sometimes the app crashes if certain errors aren’t handled correctly. It’s not critical at the moment, so we can comment it out for now.

LaurianeOSE reviewed (commented):
Here are the most important comments to address, the rest is about code style and cleaning - not important for deploying - we can address them later

notagilbert committed:
fix: remove health check route

notagilbert committed:
Merge wip/pull-6-lauriane branch and resolve migration conflicts

notagilbert committed:
Fix: Move GitHub OAuth tokens to app_user table

- Update migration 14 to add token fields to app_user instead of github_owner
- Add GitHub token methods to User repository with encryption
- Remove token methods from Owner repository
- Update controllers and strategies to use User repository for tokens
- Fix migration 15 constraints to match actual table structure
